feat(releases): Use routing for Releases Drawer

billyvg opened the pull request:
Note: this does not currently handle charts inside of the drawer list.

This splits `<ReleasesDrawer>` into `useReleasesDrawer()` hook which opens drawer and renders `<ReleasesDrawerList>` or `<ReleasesDrawerDetails>`.

billyvg committed:
feat(releases): Use routing for Releases Drawer

Note: this does not currently handle charts inside of the drawer list.

billyvg commented on static/app/views/releases/drawer/useReleasesDrawer.tsx:
```diff
@@ -0,0 +1,145 @@
+    if (rdChart) {
+      isRenderedFromClick.current = true;
+
+      // `rdChart` is a temp hack so that we can render the charts onClick by
+      // passing the render function to openDrawer, while at the same time
+      // writing a URL entry that gets ignored by `useReleasesDrawer` due to
+      // `rdChart` being present in the URL params. This hook then removes
+      // `rdChart` so that on reload, this hook will render the drawer (without
+      // chart) due to the URL params. Note that it might be possible (but
+      // unprobable?) that this `rdChart` clearing call gets interrupted and
+      // the user copies a URL with the param, in which case, the drawer won't
+      // open on next load.
+      navigate(
+        {
+          query: queryWithoutChart,
+        },
+        {replace: true}
+      );
```
🙈

billyvg commented on static/app/views/releases/releaseBubbles/useReleaseBubbles.tsx:
```diff
@@ -56,18 +73,19 @@ const trackLegend = debounce((params: LegendSelectChangedParams) => {
-  };
-  releases: ReleaseMetaBasic[];
-  theme: Theme;
+const RELEASES_DRAWER_FIELD_MAP = {
+  showReleasesDrawer: decodeScalar,
+  rdEnd: decodeScalar,
+  rdStart: decodeScalar,
+  rdEnv: decodeList,
+  rdProject: decodeList,
+  release: decodeScalar,
+  releaseProjectId: decodeScalar,
+};
+const RELEASES_DRAWER_FIELDS = Object.keys(RELEASES_DRAWER_FIELD_MAP);
+
+function cleanLocationQuery(query: Record<string, string[] | string | null | undefined>) {
+  return omit(query, RELEASES_DRAWER_FIELDS);
```
This is duped but should be temporary.

billyvg commented on static/app/views/releases/releaseBubbles/useReleaseBubbles.tsx:
```diff
@@ -441,22 +461,39 @@ export function useReleaseBubbles({
+        navigate({
+          query: {
+            ...location.query,
+            showReleasesDrawer: 1,
+            rdChart: 1,
+            rdStart: data.start,
+            rdEnd: data.end,
+            rdProject: projects ?? selection.projects,
+            rdEnv: environments ?? selection.environments,
```
thoughts on param names?

leeandher commented on static/app/components/globalDrawer/index.tsx:
```diff
@@ -130,7 +130,11 @@ export function GlobalDrawer({children}: any) {
     () => {
       // Defaults to closing the drawer when the location changes
-      if (currentDrawerConfig?.options.shouldCloseOnLocationChange?.(location) ?? true) {
+      if (
+        // Drawer should be closed already if `currentDrawerConfig` is undefined
+        currentDrawerConfig !== undefined &&
+        (currentDrawerConfig?.options.shouldCloseOnLocationChange?.(location) ?? true)
+      ) {
```
@billyvg  I believe so, but you can be explicit with `isDrawerOpen` 👍

billyvg commented:
The chart rendering hack breaks the breadcrumb navigation from release details. Going to revert this back and wait until we get chart deeplinking.

billyvg committed:
subPageFilters -> pageFilters

billyvg commented:
Do not merge until Insights + EventGraph widgets have all been refactored.

billyvg committed:
double useReleasesDrawer

billyvg committed:
update subPageFilters

billyvg committed:
Merge branch 'master' into billy/replay-2-releases-drawer-needs-to-have-deeplinks

aliu39 commented on static/app/views/releases/drawer/releasesDrawer.tsx:
```diff
@@ -1,157 +1,89 @@
-type ListProps = ReleasesDrawerListProps;
-type DetailsProps = {release: string};
-type ReleasesDrawerProps = XOR<ListProps, DetailsProps>;
+const RELEASES_DRAWER_FIELD_MAP = {
```
Could this be merged /shared with the other map in `useReleasesDrawer` maybe?

aliu39 reviewed (approved):
Looks good, although I don't think I'm familiar enough w/this area to know about missed spots or edge cases

michellewzhang commented on static/app/views/releases/drawer/releasesDrawerList.tsx:
```diff
@@ -137,54 +109,39 @@ export function ReleasesDrawerList({
+      </EventNavigator>
+      <EventDrawerBody>
+        {chart ? (
+          <ChartContainer style={{height: chartHeight}}>
+            <ChartWidgetLoader
+              id={chart}
+              pageFilters={pageFilters}
+              showReleaseAs="line"
+            />
+          </ChartContainer>
+        ) : null}
+
+        <ReleaseDrawerTable
+          {...pageFilters}
```
<img width="698" alt="SCR-20250422-naml" src="https://github.com/user-attachments/assets/053d76b7-8e10-4e65-93c9-3bc2f41a9ebf" />

![SCR-20250422-nbch](https://github.com/user-attachments/assets/624a5697-7fb5-4543-a676-48d3d88d18c9)

might need some padding here

michellewzhang commented:
any way to prevent the background charts from refetching every time a release bubble is clicked? 

https://github.com/user-attachments/assets/6f8a728f-edbd-4b86-af43-5db24de61a7c

also happens when the table is paginated:

https://github.com/user-attachments/assets/81ec7257-f8fa-49f6-971b-fdb321ae5fd2

michellewzhang commented on static/app/views/releases/drawer/releasesDrawerTable.tsx:
```diff
@@ -79,16 +73,8 @@ export function ReleaseDrawerTable({
-            start,
-            end,
-          }),
+          cursor: location.query.rdListCursor,
```
we should probably reset this cursor every time the drawer is closed -- right now it's keeping the pagination state every time a new bubble is clicked

billyvg replied:
fixed

michellewzhang commented on static/app/views/releases/drawer/releasesDrawerDetails.tsx:
```diff
@@ -23,52 +41,113 @@ export function ReleasesDrawerDetails({release, projectId}: ReleasesDrawerDetail
+            <NewIssues projectId={projectId} release={release} />
+          </FoldSection>
+        </div>
+      </EventDrawerBody>
```
https://github.com/user-attachments/assets/29112a9b-acf3-48fe-b74c-726cc42e016f

this is not related and low priority, but could we make use of memoization here or maybe increased staletime if applicable? so it doesn't show the placeholder every single time the section is collapsed/folded

billyvg replied:
[ticketed](https://linear.app/getsentry/issue/REPLAY-194/release-drawer-details-new-issues-fetches-too-often)

michellewzhang commented on static/app/views/releases/drawer/releasesDrawerDetails.tsx:
```diff
@@ -23,52 +41,113 @@ export function ReleasesDrawerDetails({release, projectId}: ReleasesDrawerDetail
+    ...locationQueryWithoutRelease
+  } = location.query;
 
   if (!projectId || !projectSlug) {
     // TODO: Error handling... not sure when this would happen
     return <Alert type="error">{t('Project not found')}</Alert>;
```
https://github.com/user-attachments/assets/896c3cd8-7235-42dd-ab5a-ba6699e67fce


it happened 💀

billyvg replied:
Partially fixed but also [added a TODO](https://linear.app/getsentry/issue/REPLAY-195/add-project-picker-to-releases-details-drawer)

billyvg committed:
Merge branch 'master' into billy/replay-2-releases-drawer-needs-to-have-deeplinks

billyvg committed:
needs to be default exports

billyvg committed:
default exports

billyvg commented:
> any way to prevent the background charts from refetching every time a release bubble is clicked?
> 
>  Screen.Recording.2025-04-22.at.2.38.57.PM.mov 
> also happens when the table is paginated:
> 
>  Screen.Recording.2025-04-22.at.2.42.51.PM.mov

Damn I didn't test in session health at all -- it looks like it's because the hooks in these charts all spread `location.query` so when we open the drawer, these hooks re-fetch due to the release drawer query params. We probably want to make these hooks use specific query params instead of spreading `location.query`.

billyvg committed:
specify height from releasedrawerlist

billyvg committed:
Merge branch 'master' into billy/replay-2-releases-drawer-needs-to-have-deeplinks

billyvg committed:
Refactor query parameters handling in the Releases Drawer components to improve maintainability and readability. This involves centralizing the query parameter logic and ensuring consistent usage across components. The changes include:

- Utilizing `RELEASES_DRAWER_FIELD_MAP` for consistent decoding of query parameters.
- Cleaning up the query parameters using `cleanLocationQuery` to remove unnecessary fields.
- Ensuring that the `useLocationQuery` hook is used effectively to manage query parameters in a centralized manner.

billyvg committed:
refactor to use query params enum

billyvg committed:
Merge branch 'master' into billy/replay-2-releases-drawer-needs-to-have-deeplinks

billyvg committed:
tweak to clear cursors only

billyvg committed:
fix wrong query params

billyvg committed:
Merge branch 'master' into billy/replay-2-releases-drawer-needs-to-have-deeplinks

billyvg committed:
temp handle case where we do not have releaseProjectId in URL

billyvg committed:
check loading state

billyvg committed:
add analytics events for releases drawer

billyvg committed:
Merge branch 'master' into billy/replay-2-releases-drawer-needs-to-have-deeplinks

billyvg committed:
fix lint

michellewzhang reviewed (approved):
loading looks good now

billyvg committed:
fix issues eventgraphwidget

billyvg committed:
fix navigation in eventGraph

billyvg committed:
more refactoring

billyvg committed:
Merge branch 'master' into billy/replay-2-releases-drawer-needs-to-have-deeplinks

billyvg committed:
lint
